File: Leetcode - Python/1005_array_largestSumAfterKNegations.py
# https://leetcode.com/problems/maximize-sum-of-array-after-k-negations/

import heapq

class Solution:
    def largestSumAfterKNegations(self, A: List[int], K: int) -> int:
        heapq.heapify(A)
        while K and A[0] < 0:
            heapq.heapreplace(A, -A[0])
            K -= 1
        
        if K % 2:
            heapq.heapreplace(A, -A[0])

        return sum(A)

File: Leetcode - Java/1005_array_largestSumAfterKNegations.java
// https://leetcode.com/problems/maximize-sum-of-array-after-k-negations/

class Solution {
    public int largestSumAfterKNegations(int[] A, int K) {
        Queue<Integer> pq = new PriorityQueue();
        for (int a : A) {
            pq.add(a);
        }
        
        while (K-- > 0) {
            pq.add(-pq.poll());
        }
        
        int sum = 0;
       for(int i = 0; i < A.length; i++){
            sum += pq.poll();
        }
        
        return sum;
    }
}